lib/log.c

```diff
@@ -83,7 +82,7 @@ void dolastlog (
 	newlog.ll_time = ll_time;
 	STRTCPY(newlog.ll_line, line);
```

Comment:
Because STRNCPY() doesn't guarantee a string, and there are a few places where we use `.ll_line` as a string:

```c
$ grep -rn ll_line -C1
lib/log.c-83-	newlog.ll_time = ll_time;
lib/log.c:84:	STRTCPY(newlog.ll_line, line);
lib/log.c-85-#if HAVE_LL_HOST
--
src/lastlog.c-172-	printf ("%-16s %-8.8s %*s%s\n",
src/lastlog.c:173:	        pw->pw_name, ll.ll_line, -maxIPv6Addrlen, ll.ll_host, cp);
src/lastlog.c-174-#else
src/lastlog.c-175-	printf ("%-16s\t%-8.8s %s\n",
src/lastlog.c:176:	        pw->pw_name, ll.ll_line, cp);
src/lastlog.c-177-#endif
--
src/lastlog.c-231-#endif
src/lastlog.c:232:		strcpy (ll.ll_line, "lastlog");
src/lastlog.c-233-#ifdef WITH_AUDIT
--
src/login.c-1205-			printf (_("Last login: %s on %s"),
src/login.c:1206:			        ptime, ll.ll_line);
src/login.c-1207-#ifdef HAVE_LL_HOST		/* __linux__ || SUN4 */
```

On the other hand, maybe we should replace those `%s` by `%.*s` and use `printf(..., UT_LINESIZE, line);`.

I'll think about it.

Regarding ignoring the return value, we want to unignore it, but didn't have time to research about it.  Hopefully, we'll do it some day.  The use of those macros was in preparation to allow us to add checks.

Reply:
BTW, I believe STRNCPY() is a useful wrapper around strncpy(3).  It prevents one from choosing a wrong value for the size of the copy, which is quite common when calling it (you should see the code I removed when added those calls).

And it's just does the obvious thing:

```c
$ grepc STRNCPY .
./lib/string/strncpy.h:#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
```

strtcpy() (called via STRTCPY()) is more complex, non-standard, and thus more controversial, but I expect STRNCPY() to be relatively uncontroversial, no?

Reply:
```
commit 225530b7e111adf7481c3e3ac172b386194c231c
Author: Alejandro Colomar <alx@kernel.org>
Date:   Wed Nov 15 22:31:02 2023 +0100

    lib/strncpy.h: Add STRNCPY() wrapper for strncpy(3)
    
    This wrapper calculates the destination buffer's size, to avoid errors
    in the size calculation.
    
    A curious fact: this macro did exist in Version 7 Unix (with a slightly
    different name).  I found it by chance, investigating the origins of
    strncpy(3) and strncat(3) in V7, after Branden suggested me to do so,
    related to recent discussions about string_copying(7).
    
            alx@debian:~/src/unix/unix/Research-V7$ grepc SCPYN .
            ./usr/src/cmd/login.c:#define SCPYN(a, b)       strncpy(a, b, sizeof(a))
    
    Our implementation is slightly better, because using nitems() we're
    protected against passing a pointer instead of an array, and it's also
    conceptually more appropriate: for wide characters, it would be
    
            #define WCSNCPY(dst, src)  wcsncpy(dst, src, NITEMS(dst))
    
    Cc: "G. Branden Robinson" <branden@debian.org>
    Signed-off-by: Alejandro Colomar <alx@kernel.org>
```

Reply:
The issue with `strncpy` is that calling it with `sizeof` as the third argument is very easily misused, mostly because arrays in C implicitly decay to pointers. Consider the following, for example:

```
: doctor; cat foo.c
#include <stdio.h>
#include <stdlib.h>

void
size(char array[20])
{
        printf("size of array = %zu\n", sizeof(array));
}

int
main(void)
{
        char arr[10];
        size(arr);
        return EXIT_SUCCESS;
}
: doctor; gcc -o foo foo.c
foo.c: In function 'size':
foo.c:7:47: warning: 'sizeof' on array function parameter 'array' will return size of 'char *' [-Wsizeof-array-argument]
    7 |         printf("size of array = %zu\n", sizeof(array));
      |                                               ^
foo.c:5:11: note: declared here
    5 | size(char array[20])
      |      ~~~~~^~~~~~~~~
foo.c: In function 'main':
foo.c:14:9: warning: 'size' accessing 20 bytes in a region of size 10 [-Wstringop-overflow=]
   14 |         size(arr);
      |         ^~~~~~~~~
foo.c:14:9: note: referencing argument 1 of type 'char[20]'
foo.c:5:1: note: in a call to function 'size'
    5 | size(char array[20])
      | ^~~~
foo.c:14:9: warning: 'size' accessing 20 bytes in a region of size 10 [-Wstringop-overflow=]
   14 |         size(arr);
      |         ^~~~~~~~~
foo.c:14:9: note: referencing argument 1 of type 'char[20]'
foo.c:5:1: note: in a call to function 'size'
    5 | size(char array[20])
      | ^~~~
: doctor; ./foo
size of array = 8
: doctor;
```

Here, at least the compiler has the decency to tell me that what I'm doing isn't what one might naively expect, but it's still really easy to get this wrong, particularly if it's hidden behind a macro.

It would be safer to always pass a symbolic constant denoting the size of the destination (in this case, `UT_LINESIZE`), IMHO.

Reply:
But we don't call sizeof() directly.

```c
alx@debian:~/src/shadow/shadow/master$ grepc STRNCPY .
./lib/string/strncpy.h:#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
alx@debian:~/src/shadow/shadow/master$ grepc NITEMS .
./lib/sizeof.h:#define NITEMS(a)            (SIZEOF_ARRAY((a)) / sizeof((a)[0]))
```

The NITEMS() macro does give us the necessary warnings to prevent what you're worried about (mentioned in the commit message I pasted above).

```c
alx@debian:~/tmp/gcc$ cat strncpy.c 
#include <string.h>

#define NITEMS(a)          (sizeof(a) / sizeof((a)[0]))
#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))

int
main(void)
{
	char  *s = "a";
	char  a[10];
	char  *p = a;

	STRNCPY(a, s);  // ok
	STRNCPY(p, s);  // not ok
}
```
```c
alx@debian:~/tmp/gcc$ gcc -Wall strncpy.c 
strncpy.c: In function ‘main’:
strncpy.c:3:39: warning: division ‘sizeof (char *) / sizeof (char)’ does not compute the number of array elements [-Wsizeof-pointer-div]
    3 | #define NITEMS(a)          (sizeof(a) / sizeof((a)[0]))
      |                                       ^
strncpy.c:4:46: note: in expansion of macro ‘NITEMS’
    4 | #define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
      |                                              ^~~~~~
strncpy.c:14:9: note: in expansion of macro ‘STRNCPY’
   14 |         STRNCPY(p, s);  // not ok
      |         ^~~~~~~
strncpy.c:11:16: note: first ‘sizeof’ operand was declared here
   11 |         char  *p = a;
      |                ^
```

Reply:
re: `STRTCPY`, I see now how it handles unconditionally terminating the destination (unless `dsize == 0`). My initial glance at gave me the impression that it was exactly the same as `strncpy` except for the return value, but I'd missed `dlen = slen - trunc` on line 69 of `strtcpy.h`. My bad!

Reply:
Plus, we actually have stronger guarantees, since NITEMS() actually uses SIZEOF_ARRAY(), which is a static_assert(3) in disguise.

```c
alx@debian:~/src/shadow/shadow/master$ grepc NITEMS .
./lib/sizeof.h:#define NITEMS(a)            (SIZEOF_ARRAY((a)) / sizeof((a)[0]))
alx@debian:~/src/shadow/shadow/master$ grepc SIZEOF_ARRAY .
./lib/sizeof.h:#define SIZEOF_ARRAY(a)      (sizeof(a) + must_be_array(a))
alx@debian:~/src/shadow/shadow/master$ grepc must_be_array .
./lib/must_be.h:#define must_be_array(a)      must_be(is_array(a))
alx@debian:~/src/shadow/shadow/master$ grepc is_array .
./lib/must_be.h:#define is_array(a)           (!is_same_typeof((a), &(a)[0]))
alx@debian:~/src/shadow/shadow/master$ grepc must_be .
./lib/must_be.h:#define must_be(e)                                                            \
(                                                                             \
	0 * (int) sizeof(                                                     \
		struct {                                                      \
			static_assert(e, "");                                 \
			int ISO_C_forbids_a_struct_with_no_members_;          \
		}                                                             \
	)                                                                     \
)
alx@debian:~/src/shadow/shadow/master$ grepc is_same_typeof .
./lib/must_be.h:#define is_same_typeof(a, b)  is_same_type(typeof(a), typeof(b))
alx@debian:~/src/shadow/shadow/master$ grepc is_same_type .
./lib/must_be.h:#define is_same_type(a, b)    __builtin_types_compatible_p(a, b)
```

Reply:
> But we don't call sizeof() directly.

I know. :-) But you get the same warning with `sizeof` by itself.

> The NITEMS() macro does give us the necessary warnings to prevent what you're worried about (mentioned in the commit message I pasted above).

I still think you're playing with fire, IMHO, but I suppose if you compile with `-Werror` and you feel confident the code has been sufficiently audited, I have no real standing to complain and defer to your judgment.

Reply:
> Plus, we actually have stronger guarantees, since NITEMS() actually uses SIZEOF_ARRAY(), which uses static_assert(3) internally.

Ah! You're good to go, then. C11 and later FTW.

Reply:
> re: `STRTCPY`, I see now how it handles unconditionally terminating the destination (unless `dsize == 0`). My initial glance at gave me the impression that it was exactly the same as `strncpy` except for the return value, but I'd missed `dlen = slen - trunc` on line 69 of `strtcpy.h`. My bad!

Yup; strtcpy() is conceptually more or less the same as Linux's strscpy(9), except that we return -1 on error but they do -E2BIG.

Reply:
> > But we don't call sizeof() directly.
> 
> I know. :-) But you get the same warning with `sizeof` by itself.

Actually, not the same one.

With NITEMS() (the sizeof division), you get `-Wsizeof-pointer-div`, but with sizeof() you get `-Wsizeof-pointer-memaccess`.

```c
$ cat strncpy.c 
#include <string.h>

#define NITEMS(a)          (sizeof(a) / sizeof((a)[0]))
#define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))

int
main(void)
{
	char  *s = "a";
	char  a[10];
	char  *p = a;

	STRNCPY(a, s);  // ok
	STRNCPY(p, s);  // not ok
	strncpy(a, s, sizeof(a));  // ok
	strncpy(p, s, sizeof(p));  // not ok
}
```
```c
$ cc -Wall strncpy.c 
strncpy.c: In function ‘main’:
strncpy.c:3:39: warning: division ‘sizeof (char *) / sizeof (char)’ does not compute the number of array elements [-Wsizeof-pointer-div]
    3 | #define NITEMS(a)          (sizeof(a) / sizeof((a)[0]))
      |                                       ^
strncpy.c:4:46: note: in expansion of macro ‘NITEMS’
    4 | #define STRNCPY(dst, src)  strncpy(dst, src, NITEMS(dst))
      |                                              ^~~~~~
strncpy.c:14:9: note: in expansion of macro ‘STRNCPY’
   14 |         STRNCPY(p, s);  // not ok
      |         ^~~~~~~
strncpy.c:11:16: note: first ‘sizeof’ operand was declared here
   11 |         char  *p = a;
      |                ^
strncpy.c:16:29: warning: argument to ‘sizeof’ in ‘strncpy’ call is the same expression as the destination; did you mean to provide an explicit length? [-Wsizeof-pointer-memaccess]
   16 |         strncpy(p, s, sizeof(p));  // not ok
      |                             ^
```

> 
> > The NITEMS() macro does give us the necessary warnings to prevent what you're worried about (mentioned in the commit message I pasted above).
> 
> I still think you're playing with fire, IMHO, but I suppose if you compile with `-Werror` and you feel confident the code has been sufficiently audited, I have no real standing to complain and defer to your judgment.

The static_assert(3) called within NITEMS() guarantees that we don't play with fire.  We don't even need -Werror, because static_assert(3) is a guaranteed compile-time error:

```c
$ cat sa.c 
static_assert(0);

int
main(void)
{
}
```
```c
$ cc sa.c 
sa.c:1:15: error: expected declaration specifiers or ‘...’ before numeric constant
    1 | static_assert(0);
      |               ^
```

See <https://github.com/shadow-maint/shadow/pull/1022#discussion_r1653774117>

Reply:
I hadn't updated the page, and hadn't seen your reply when I sent this.  Sorry.  We agree.  :)

Reply:
(Tangentially to this, one wonders if a different name than `STRNCPY` would be more appropriate; perhaps `SARRCPY`; nothing in the definition of `strncpy` requires that it's destination argument be an array, but the `STRNCPY` macro does. This is much of the source of my initial confusion here; well, that and not taking the time to peel back the several layers of macros to get to the static assert.... This is well outside of the scope of this review, however.)

Reply:
> (Tangentially to this, one wonders if a different name than `STRNCPY` would be more appropriate; perhaps `SARRCPY`; nothing in the definition of `strncpy` requires that it's destination argument be an array, but the `STRNCPY` macro does. This is much of the source of my initial confusion here; well, that and not taking the time to peel back the several layers of macros to get to the static assert.... This is well outside of the scope of this review, however.)

I'm happy to discuss this macro name here.  :)

Theoretically, nothing in strncpy(3) requires that the destination is an array.  However, I think that the only _appropriate_ situation in which strncpy(3) _should_ be called is with a destination array.  In every other situation, there's a more appropriate call.  See string_copying(7) for a detailed document.

Reply:
> > (Tangentially to this, one wonders if a different name than `STRNCPY` would be more appropriate; perhaps `SARRCPY`; nothing in the definition of `strncpy` requires that it's destination argument be an array, but the `STRNCPY` macro does. This is much of the source of my initial confusion here; well, that and not taking the time to peel back the several layers of macros to get to the static assert.... This is well outside of the scope of this review, however.)
> 
> I'm happy to discuss this macro name here. :)
> 
> Theoretically, nothing in strncpy(3) requires that the destination is an array. However, I think that the only _appropriate_ situation in which strncpy(3) _should_ be called is with a destination array. In every other situation, there's a more appropriate call.

Well, now we're getting into the history of the thing (https://groups.io/g/multicians/message/2571; note my mistake: `strncpy` didn't originate in the kernel: https://www.tuhs.org/mailman3/hyperkitty/list/tuhs@tuhs.org/message/ZND7BO4SH7RQTAYYDH62J7A7W2PSIEF5/).

`strncpy` has arguably been misnamed from the start, but its semantics have been more or less well-defined for decades (if often misunderstood). If introducing a new macro with different semantics, I would argue that a different name is appropriate. After all, character fields can be referenced by pointer a la `strncpy`, but not so with `STRNCPY`.

> See string_copying(7) for a detailed document.

Wow. That's an opinionated document.... :-)

Reply:
> > > (Tangentially to this, one wonders if a different name than `STRNCPY` would be more appropriate; perhaps `SARRCPY`; nothing in the definition of `strncpy` requires that it's destination argument be an array, but the `STRNCPY` macro does. This is much of the source of my initial confusion here; well, that and not taking the time to peel back the several layers of macros to get to the static assert.... This is well outside of the scope of this review, however.)
> > 
> > 
> > I'm happy to discuss this macro name here. :)
> > Theoretically, nothing in strncpy(3) requires that the destination is an array. However, I think that the only _appropriate_ situation in which strncpy(3) _should_ be called is with a destination array. In every other situation, there's a more appropriate call.
> 
> Well, now we're getting into the history of the thing (https://groups.io/g/multicians/message/2571;

I have applied to that group now (with `<alx.manpages@gmail.com>`).  I need to be approved before being allowed to read that.

> note my mistake: `strncpy` didn't originate in the kernel: https://www.tuhs.org/mailman3/hyperkitty/list/tuhs@tuhs.org/message/ZND7BO4SH7RQTAYYDH62J7A7W2PSIEF5/).
> 
> `strncpy` has arguably been misnamed from the start,

Indeed.

> After all, character fields can be referenced by pointer a la `strncpy`, but not so with `STRNCPY`.

I guess you mean

```c
*stpncpy(dst, arr, NITEMS(arr)) = '\0';
// or equivalently:
stpcpy(stpncpy(dst, arr, NITEMS(arr)), "");  // This protects the last byte with _FORTIFY_SOURCE
```

(or the equivalent with strncpy(3).  I didn't even write the strncpy(3) equivalent because I know I'll get it off-by-one if I write it.)

Am I correct in that assumption?

That's true.  In string_copying(7), I recommend using the following instead:

```c
stpcpy(mempcpy(dst, arr, strnlen(arr, NITEMS(arr))), "");
```

which is only slightly more code.  This is semantically better, since it uses stpcpy(mempcpy()), which denotes that the result is a string, and uses strnlen(NITEMS()) on the input, which denotes that it's a null-padded character sequence.  It also omits the unnecessary and confusing zeroing.

We have a macro for doing that operation: we call it ZUSTR2STP().  And it does the necessary static assertions for the operation:

```c
$ grepc ZUSTR2STP .
./lib/string/zustr2stp.h:#define ZUSTR2STP(dst, src)                                                   \
({                                                                            \
	static_assert(!is_array(dst) || sizeof(dst) > SIZEOF_ARRAY(src), ""); \
                                                                              \
	stpcpy(mempcpy(dst, src, strnlen(src, NITEMS(src))), "");             \
})
```

ZUSTR is because it's a Zero-padded Unterminated STRing.
2 because it goes from a ZUSTR to a STP.
STP because it oproduces a string, and returns a pointer to the NUL, like stpcpy(3).

> If introducing a new macro with different semantics, I would argue that a different name is appropriate.

As you can see in string_copying(7), my plan is to reduce the semantics of strncpy(3) to only one of the existing use cases, via documentation, and consider the rest as things to avoid.  STRNCPY() is a macro that works for that intended semantics and refuses to compile with the other deprecated semantics.  I consider that a bugfix, not a change.

> but its semantics have been more or less well-defined for decades (if often misunderstood).

I'm not fully agreeing on that.  When I see *strncpy*(3) misused for something that is not "filling a fixed-size buffer with characters from a string and pad with null bytes", I have a hard time determining if the zeroing is intentional or just a consequence of abusing this API.  So, the semantics are actually quite confusing.  I also often see calls with off-by-one errors (stpncpy(3) solves some of those).

I'm even unhappy with out ZUSTR2STP() (even if it's way better than strncpy(3)), and have a patch set to replace it by calls to strndup(3) (or actually, xstrndup()) via a XSTRNDUP() macro (the X because we abort on ENOMEM):

<https://github.com/shadow-maint/shadow/pull/991/commits/c80dee371965819eb107440342e6beaff370fb81>
<https://github.com/shadow-maint/shadow/pull/991/commits/450e41457d5ac05b0717d410ff473544f37906d9>
<https://github.com/shadow-maint/shadow/pull/991/commits/ee8390d1c62e58cfd9c7c3236475a66fc3b86d06>

> > See string_copying(7) for a detailed document.
> 
> Wow. That's an opinionated document.... :-)

It is.  I hope it serves to improve the long-standing misuse of these functions.  :-)

Reply:
Oh, and that patch set reminds me that there's the obvious `strncat(stpcpy(dst, ""), src, NITEMS(src));` to supersede that bogus use of strncpy(3).

Reply:
> > > > (Tangentially to this, one wonders if a different name than `STRNCPY` would be more appropriate; perhaps `SARRCPY`; nothing in the definition of `strncpy` requires that it's destination argument be an array, but the `STRNCPY` macro does. This is much of the source of my initial confusion here; well, that and not taking the time to peel back the several layers of macros to get to the static assert.... This is well outside of the scope of this review, however.)
> > > 
> > > 
> > > I'm happy to discuss this macro name here. :)
> > > Theoretically, nothing in strncpy(3) requires that the destination is an array. However, I think that the only _appropriate_ situation in which strncpy(3) _should_ be called is with a destination array. In every other situation, there's a more appropriate call.
> > 
> > 
> > Well, now we're getting into the history of the thing (https://groups.io/g/multicians/message/2571;
> 
> I have applied to that group now (with `<alx.manpages@gmail.com>`). I need to be approved before being allowed to read that.

Eh, it wasn't _that_ interesting. I forwarded it to you.

> > note my mistake: `strncpy` didn't originate in the kernel: https://www.tuhs.org/mailman3/hyperkitty/list/tuhs@tuhs.org/message/ZND7BO4SH7RQTAYYDH62J7A7W2PSIEF5/).
> > `strncpy` has arguably been misnamed from the start,
> 
> Indeed.
> 
> > After all, character fields can be referenced by pointer a la `strncpy`, but not so with `STRNCPY`.
> 
> I guess you mean
> [snip]
> (or the equivalent with strncpy(3). I didn't even write the strncpy(3) equivalent because I know I'll get it off-by-one if I write it.)
> 
> Am I correct in that assumption?

What I mean is that, unfortunately, there are a number of use cases in which `strncpy` _can_ be used where the destination argument definition as an array is not directly accessible to be passed to `NITEMS` (or some equivalent). Perhaps the obvious one is when the destination is a malloc'd array, in which case `NITEMS` won't do the right thing. Others include when the destination is passed indirectly, as in via an argument to some intermediary function that invokes strncpy, but again without access to the original destination array definition (that was my initial example). Or perhaps where the argument is returned by some other function, such as some accessor function that returns a pointer to a static array, but that the caller expects to modify. Or, consider the case where the destination argument _is_ an array, but not an array of characters; perhaps one wants to stuff character data into a machine register, for example, or something that looks like a machine register. An example here might be filling in the CPUID leaf for a CPU's version string in a hypervisor; on x86, the destination is a de facto character field, but it's represented as a sequence of uint32_t's that are filled into registers in a VMEXIT handler.

E.g.,

```c
    uint32_t arr[4];
    strncpy((char *)arr, "blahblahblah1", sizeof(arr));
````

One might validly argument that any of these are poor fits for `strncpy`, and that may well be true. But nonetheless, they exist.

Anyway, I'm not really trying to change your opinion (which I largely agree with), but I do want to point out that there's ~45 years of history there, and the legacy of that is not going away soon.

> That's true. In string_copying(7), I recommend using the following instead:
> 
> ```c
> stpcpy(mempcpy(dst, arr, strnlen(arr, NITEMS(arr))), "");
> ```
> 
> which is only slightly more code. This is semantically better, since it uses stpcpy(mempcpy()), which denotes that the result is a string, and uses strnlen(NITEMS()) on the input, which denotes that it's a null-padded character sequence. It also omits the unnecessary and confusing zeroing.

I don't see how `strnlen` a priori denotes NUL-padding. Consider an array such as,

`char buf[] = { 1, 1, 1, 0, 1, 1, 1, 1 };`

Here, `NITEMS(buf)` is 8, but `strnlen(buf, NITEMS(buf))` is 3, and yet `buf` is not NUL-padded; it just contains a single NUL character in the 4th element. I do, however, take your point that `strnlen` indicates that one expects that a NUL _could_ appear somewhere in the $n$ bytes of it's first argument.

I know I'm opening a can of worms here, but why not just import `strlcpy` and be done with it? It seems to do exactly what you want, but with less complexity and more widespread adoption.

> We have a macro for doing that operation: we call it ZUSTR2STP(). And it does the necessary static assertions for the operation:
> 
> ```c
> $ grepc ZUSTR2STP .
> ./lib/string/zustr2stp.h:#define ZUSTR2STP(dst, src)                                                   \
> ({                                                                            \
> 	static_assert(!is_array(dst) || sizeof(dst) > SIZEOF_ARRAY(src), ""); \
>                                                                               \
> 	stpcpy(mempcpy(dst, src, strnlen(src, NITEMS(src))), "");             \
> })
> ```
> 
> ZUSTR is because it's a Zero-padded Unterminated STRing. 2 because it goes from a ZUSTR to a STP. STP because it oproduces a string, and returns a pointer to the NUL, like stpcpy(3).

This, I think, strengthens my argument about the macro name. Indeed, I view this as something of an opportunity: `strncpy` has been misnamed since its inception, its semantics are weird and unintuitive, and as a result, it has tripped a lot of programmers, resulting in very serious bugs, over a very long time.

Here, you've got the chance to assign a new name to a new macro with more rational semantics. It's an opportunity! Why not seize it?

> > If introducing a new macro with different semantics, I would argue that a different name is appropriate.
> 
> As you can see in string_copying(7), my plan is to reduce the semantics of strncpy(3) to only one of the existing use cases, via documentation, and consider the rest as things to avoid. STRNCPY() is a macro that works for that intended semantics and refuses to compile with the other deprecated semantics. I consider that a bugfix, not a change: STRNCPY() is a strncpy(3) variant that prevents misuse.

The universe of existing, useful, userspace software is very large. The pitfalls of `strncpy` are well-known, but this is the universe we've come to exist in. Bluntly, there's a lot of code out there that uses `strncpy` in dubious ways, and it's not going to change easily. It'd be better to pick a new name for something with new semantics. As an outsider (I don't usually use Linux, even) I saw that macro and got confused as to where the third argument was; then I got confused as to what happens if one passes a pointer to it for the destination argument. I'm not the smartest person on the planet, but I can say that, for me personally, a different name probably would have reduced if not eliminated most of that confusion.

> > but its semantics have been more or less well-defined for decades (if often misunderstood).
> 
> I'm not fully agreeing on that. When I see _strncpy_(3) misused for something that is not "filling a fixed-size buffer with characters from a string and pad with null bytes", I have a hard time determining if the zeroing is intentional or just a consequence of abusing this API. So, the semantics are actually quite confusing. I also often see calls with off-by-one errors (stpncpy(3) solves some of those).

No, the _semantics of `strncpy`_ are very well defined in the C standard and POSIX, etc. Whether it's _used semantically_ in any given case is another matter, and that I agree is often difficult to tell; in my experience, it is almost always misused. For example, when I worked on the Akaros research OS, I audited all uses of `strncpy` in the kernel; of however-many sites, I determined that 4 were correct; the rest were just wrong. But that's not _really_ `strncpy`'s fault; in this case, that was the fault of programmers who used Coccinelle to automatically modify a bunch of code imported from plan9, replacing calls to `strcpy` (which we excised entirely).

> I'm even unhappy with our ZUSTR2STP() (even if it's way better than strncpy(3)), and have a patch set to replace it by calls to strndup(3) (or actually, xstrndup()) via a XSTRNDUP() macro (the X because we abort on ENOMEM):
> 
> #991 [51dc60f](https://github.com/shadow-maint/shadow/commit/51dc60f74bc6e30f93c092c347aa797bc1989d59) [c80dee3](https://github.com/shadow-maint/shadow/commit/c80dee371965819eb107440342e6beaff370fb81) [450e414](https://github.com/shadow-maint/shadow/commit/450e41457d5ac05b0717d410ff473544f37906d9) [ee8390d](https://github.com/shadow-maint/shadow/commit/ee8390d1c62e58cfd9c7c3236475a66fc3b86d06) [bf50bae](https://github.com/shadow-maint/shadow/commit/bf50bae2480134985cc5acbdeb8859bf65ba0d8d) [bf50bae](https://github.com/shadow-maint/shadow/commit/bf50bae2480134985cc5acbdeb8859bf65ba0d8d)

That seems like a net improvement.

> > > See string_copying(7) for a detailed document.
> > 
> > 
> > Wow. That's an opinionated document.... :-)
> 
> It is. I hope it serves to improve the long-standing misuse of these functions. :-)

One can hope! To circle back to the origin of this thread, I worry about very old code that's more or less unmaintained, but still somewhat used. Getting that updated is always going to be an uphill slog. :-( Though of course we shouldn't halt the march of progress for it....

Reply:
> > > After all, character fields can be referenced by pointer a la `strncpy`, but not so with `STRNCPY`.
> > 
> > 
> > I guess you mean
> > [snip]
> > (or the equivalent with strncpy(3). I didn't even write the strncpy(3) equivalent because I know I'll get it off-by-one if I write it.)
> > Am I correct in that assumption?
> 
> What I mean is that, unfortunately, there are a number of use cases in which `strncpy` _can_ be used where the destination argument definition as an array is not directly accessible to be passed to `NITEMS` (or some equivalent).

> Perhaps the obvious one is when the destination is a malloc'd array, in which case `NITEMS` won't do the right thing.

When malloc'ing memory, you're free to allocate one more byte, so I don't see why one'd want to make it a fixed-fidth buffer that contains a length-bounded character sequence instead of a string.

Except if the malloc'd memory is a structure with a specific layout.  But if it's a structure, the struct will have an array member, which will allow the use of STRNCPY().

> Others include when the destination is passed indirectly, as in via an argument to some intermediary function that invokes strncpy, but again without access to the original destination array definition (that was my initial example).

Since those arrays are usually within structures, you get the full array, and not just a pointer.  I have yet to meet a case where the array is decayed (I guess it's such a niche use case that happens seldom-to-never).  But I conceive it to be possible.  In such a case, I would consider some workaround, such as re-implementing that function as a macro, or refactoring in a way that I get the true array.

> Or perhaps where the argument is returned by some other function, such as some accessor function that returns a pointer to a static array, but that the caller expects to modify. Or, consider the case where the destination argument _is_ an array, but not an array of characters; perhaps one wants to stuff character data into a machine register, for example, or something that looks like a machine register. An example here might be filling in the CPUID leaf for a CPU's version string in a hypervisor; on x86, the destination is a de facto character field, but it's represented as a sequence of uint32_t's that are filled into registers in a VMEXIT handler.
> 
> E.g.,
> 
> ```c
>     uint32_t arr[4];
>     strncpy((char *)arr, "blahblahblah1", sizeof(arr));
> ```
> 
> One might validly argument that any of these are poor fits for `strncpy`, and that may well be true. But nonetheless, they exist.

I guess there do exist some cases that need strncpy(3).

> 
> Anyway, I'm not really trying to change your opinion (which I largely agree with), but I do want to point out that there's ~45 years of history there, and the legacy of that is not going away soon.
> 
> > That's true. In string_copying(7), I recommend using the following instead:
> > ```c
> > stpcpy(mempcpy(dst, arr, strnlen(arr, NITEMS(arr))), "");
> > ```
> > 
> > which is only slightly more code. This is semantically better, since it uses stpcpy(mempcpy()), which denotes that the result is a string, and uses strnlen(NITEMS()) on the input, which denotes that it's a null-padded character sequence. It also omits the unnecessary and confusing zeroing.
> 
> I don't see how `strnlen` a priori denotes NUL-padding.

It doesn't  by itself, but it does in this context...

> Consider an array such as,
> 
> `char buf[] = { 1, 1, 1, 0, 1, 1, 1, 1 };`
> 
> Here, `NITEMS(buf)` is 8, but `strnlen(buf, NITEMS(buf))` is 3, and yet `buf` is not NUL-padded; it just contains a single NUL character in the 4th element. I do, however, take your point that `strnlen` indicates that one expects that a NUL _could_ appear somewhere in the n bytes of it's first argument.

... I think we agree.

> 
> I know I'm opening a can of worms here, but why not just import `strlcpy` and be done with it? It seems to do exactly what you want, but with less complexity and more widespread adoption.

strlcpy(3) truncates a byte too much (in the corner case where there's no size for the NUL):

```c
char buf[4];
strncpy(buf, "asdf", 4);  // results in {'a', 's', 'd', 'f'}
strlcpy(buf, "asdf", 4);  // results in "asd"
```

> 
> > We have a macro for doing that operation: we call it ZUSTR2STP(). And it does the necessary static assertions for the operation:
> > ```c
> > $ grepc ZUSTR2STP .
> > ./lib/string/zustr2stp.h:#define ZUSTR2STP(dst, src)                                                   \
> > ({                                                                            \
> > 	static_assert(!is_array(dst) || sizeof(dst) > SIZEOF_ARRAY(src), ""); \
> >                                                                               \
> > 	stpcpy(mempcpy(dst, src, strnlen(src, NITEMS(src))), "");             \
> > })
> > ```
> > 
> > 
> > ZUSTR is because it's a Zero-padded Unterminated STRing. 2 because it goes from a ZUSTR to a STP. STP because it oproduces a string, and returns a pointer to the NUL, like stpcpy(3).
> 
> This, I think, strengthens my argument about the macro name. Indeed, I view this as something of an opportunity: `strncpy` has been misnamed since its inception, its semantics are weird and unintuitive, and as a result, it has tripped a lot of programmers, resulting in very serious bugs, over a very long time.

> 
> Here, you've got the chance to assign a new name to a new macro with more rational semantics. It's an opportunity! Why not seize it?

I considered it back then; and almost did it.  However, many people complained that they didn't understand the new names, so I considered only renaming when the new API wasn't just a thin wrapper around a single libc function.

I had in mind the name str2zustr() for strncpy(), since it reads a string and writes a zero-padded unterminated string.

Since STRNCPY() gets it's name after an ISO C function, and modulo the size calculation does the same byte operations, people seem to understand it.  Also, there's some tradition in this project of using upper-name macros named as libc functions when all they do is add NITEMS() calculations or other obvious stuff.  We have MALLOC() for example, which is malloc(3) with some trivial calculations inside.

```c
$ grepc MALLOC .
./lib/alloc.h:#define MALLOC(n, type)   ((type *) mallocarray(n, sizeof(type)))
$ grepc -tfd mallocarray .
./lib/alloc.h:inline void *
mallocarray(size_t nmemb, size_t size)
{
	return reallocarray(NULL, nmemb, size);
}
```

(If you're curious, look at the commit messages that added and modified those interfaces.  It's one of the APIs I'm most proud of, together with stpeprintf().)

There was also STRLCPY() in the past, when we used strlcpy(3) temporarily, which passed the obvious size internally, and other similar cases.

> > > If introducing a new macro with different semantics, I would argue that a different name is appropriate.
> > 
> > 
> > As you can see in string_copying(7), my plan is to reduce the semantics of strncpy(3) to only one of the existing use cases, via documentation, and consider the rest as things to avoid. STRNCPY() is a macro that works for that intended semantics and refuses to compile with the other deprecated semantics. I consider that a bugfix, not a change: STRNCPY() is a strncpy(3) variant that prevents misuse.
> 
> The universe of existing, useful, userspace software is very large. The pitfalls of `strncpy` are well-known, but this is the universe we've come to exist in. Bluntly, there's a lot of code out there that uses `strncpy` in dubious ways, and it's not going to change easily. It'd be better to pick a new name for something with new semantics. As an outsider (I don't usually use Linux, even) I saw that macro and got confused as to where the third argument was; then I got confused as to what happens if one passes a pointer to it for the destination argument. I'm not the smartest person on the planet, but I can say that, for me personally, a different name probably would have reduced if not eliminated most of that confusion.

I see your point.

But I only intend this macro name to be useful in this project, and maybe ar(1) and tar(1).  Not something to be released in libc for the general public.  So the public that will be surprised by it is relatively small.

> > > but its semantics have been more or less well-defined for decades (if often misunderstood).
> > 
> > 
> > I'm not fully agreeing on that. When I see _strncpy_(3) misused for something that is not "filling a fixed-size buffer with characters from a string and pad with null bytes", I have a hard time determining if the zeroing is intentional or just a consequence of abusing this API. So, the semantics are actually quite confusing. I also often see calls with off-by-one errors (stpncpy(3) solves some of those).
> 
> No, the _semantics of `strncpy`_ are very well defined in the C standard and POSIX, etc. Whether it's _used semantically_ in any given case is another matter, and that I agree is often difficult to tell; in my experience, it is almost always misused. For example, when I worked on the Akaros research OS, I audited all uses of `strncpy` in the kernel; of however-many sites, I determined that 4 were correct; the rest were just wrong. But that's not _really_ `strncpy`'s fault; in this case, that was the fault of programmers who used Coccinelle to automatically modify a bunch of code imported from plan9, replacing calls to `strcpy` (which we excised entirely).

Agree.  I don't intend those to use STRNCPY(), but rather go back to strcpy(3), or strlcpy(3).  Thus this is largely irrelevant to the large amount of strncpy(3) (mis)users.

> 
> > I'm even unhappy with our ZUSTR2STP() (even if it's way better than strncpy(3)), and have a patch set to replace it by calls to strndup(3) (or actually, xstrndup()) via a XSTRNDUP() macro (the X because we abort on ENOMEM):
> > #991 [51dc60f](https://github.com/shadow-maint/shadow/commit/51dc60f74bc6e30f93c092c347aa797bc1989d59) [c80dee3](https://github.com/shadow-maint/shadow/commit/c80dee371965819eb107440342e6beaff370fb81) [450e414](https://github.com/shadow-maint/shadow/commit/450e41457d5ac05b0717d410ff473544f37906d9) [ee8390d](https://github.com/shadow-maint/shadow/commit/ee8390d1c62e58cfd9c7c3236475a66fc3b86d06) [bf50bae](https://github.com/shadow-maint/shadow/commit/bf50bae2480134985cc5acbdeb8859bf65ba0d8d) [bf50bae](https://github.com/shadow-maint/shadow/commit/bf50bae2480134985cc5acbdeb8859bf65ba0d8d)
> 
> That seems like a net improvement.

Thanks!

> 
> > > > See string_copying(7) for a detailed document.
> > > 
> > > 
> > > Wow. That's an opinionated document.... :-)
> > 
> > 
> > It is. I hope it serves to improve the long-standing misuse of these functions. :-)
> 
> One can hope! To circle back to the origin of this thread, I worry about very old code that's more or less unmaintained, but still somewhat used. Getting that updated is always going to be an uphill slog. :-( Though of course we shouldn't halt the march of progress for it....

I wouldn't recommend that those use STRNCPY(), not even suggest that this macro should be in libc, so I guess it's fine here.

Reply:
Here are pointers to the main commits, to simplify the search for you:

<https://github.com/shadow-maint/shadow/commit/d81506de1e8e2ba544a30e54e863fcafda9cde86>
<https://github.com/shadow-maint/shadow/commit/6e58c1275252f3314d1aa5cc4d7e7f9068e3a902>
<https://github.com/shadow-maint/shadow/commit/efbbcade43ff2dca2b7a271dcbd186be08ac1913>
<https://github.com/shadow-maint/shadow/commit/09775d3718df216c75b62d73bbcc5aa060e0fc94>

And talking about innovative APIs that fix long-standing broken (or semi-broken) libc APIs, there's also liba2i, the development of which helped find and fix a bug in NetBSD's strtoi(3), and which has also helped fix uncountable bugs in shadow.  A very recommendable API.  This one, I've provided it in a standalone library intended for the general public:

<https://git.kernel.org/pub/scm/libs/liba2i/liba2i.git/>
<https://gnats.netbsd.org/cgi-bin/query-pr-single.pl?number=57828>

Have a lovely day, Dan!